extensions/incubator/src/main/java/io/opentelemetry/extension/incubator/trace/ExtendedTracer.java

```diff
@@ -5,54 +5,96 @@
 
 package io.opentelemetry.extension.incubator.trace;
 
-import io.opentelemetry.api.trace.Span;
-import io.opentelemetry.api.trace.SpanBuilder;
+import io.opentelemetry.api.baggage.Baggage;
+import io.opentelemetry.api.baggage.BaggageBuilder;
 import io.opentelemetry.api.trace.Tracer;
+import io.opentelemetry.context.Context;
 import io.opentelemetry.context.Scope;
-import java.util.concurrent.Callable;
+import java.util.Map;
 
-/** Provides easy mechanisms for wrapping standard Java constructs with an OpenTelemetry Span. */
+/**
+ * Utility class to simplify tracing.
+ *
+ * <p>The <a
+ * href="https://github.com/open-telemetry/opentelemetry-java-contrib/blob/main/extended-tracer/README.md">README</a>
+ * explains the use cases in more detail.
+ */
 public final class ExtendedTracer implements Tracer {
 
   private final Tracer delegate;
 
-  /** Create a new {@link ExtendedTracer} that wraps the provided Tracer. */
+  private ExtendedTracer(Tracer delegate) {
+    this.delegate = delegate;
+  }
+
+  /**
+   * Creates a new instance of {@link ExtendedTracer}.
+   *
+   * @param delegate the {@link Tracer} to use
+   */
   public static ExtendedTracer create(Tracer delegate) {
     return new ExtendedTracer(delegate);
   }
 
-  private ExtendedTracer(Tracer delegate) {
-    this.delegate = delegate;
+  /**
+   * Creates a new {@link ExtendedSpanBuilder} with the given span name.
+   *
+   * @param spanName the name of the span
+   * @return the {@link ExtendedSpanBuilder}
+   */
+  @Override
+  public ExtendedSpanBuilder spanBuilder(String spanName) {
+    return new ExtendedSpanBuilder(delegate.spanBuilder(spanName));
   }
 
-  /** Run the provided {@link Runnable} and wrap with a {@link Span} with the provided name. */
-  public void run(String spanName, Runnable runnable) {
-    Span span = delegate.spanBuilder(spanName).startSpan();
-    try (Scope scope = span.makeCurrent()) {
-      runnable.run();
-    } catch (Throwable e) {
-      span.recordException(e);
-      throw e;
-    } finally {
-      span.end();
-    }
+  /**
+   * Runs the given {@link SpanRunnable} in a new span with the given name and with kind INTERNAL.
+   *
+   * <p>If an exception is thrown by the {@link SpanRunnable}, the span will be marked as error, and
+   * the exception will be recorded.
+   *
+   * @param spanName the name of the span
+   * @param runnable the {@link SpanRunnable} to run
+   * @param <E> the type of the exception
+   */
+  public <E extends Throwable> void run(String spanName, SpanRunnable<E> runnable) throws E {
```

Comment:
These methods seem like unnecessary semantic sugar now.

Reply:
I was also undecided - I'll remove it.